File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
private:
    bool solve(int ind, int target, vector<int>& nums, vector<vector<int>>& dp){
        if(target == 0) return true;
        if(ind == 0) return (nums[0] == target);
        if(dp[ind][target] != -1) return dp[ind][target];
        bool exclude = solve(ind-1, target, nums, dp);
        bool include = false;
        if(target >= nums[ind]) include = solve(ind-1, target-nums[ind], nums, dp);

        return dp[ind][target] = include | exclude;
    }
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0, n = nums.size();
        for(auto num : nums) sum += num;
        if(sum % 2 != 0) return false;
        vector<vector<int>> dp(n, vector<int>((sum/2)+1, -1));
        return solve(n-1, sum/2, nums, dp);
    }
};

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    private boolean isSubsetSum(int[] arr, int sum){
        int n = arr.length;
        boolean[][] dp = new boolean[n][sum+1];
        for(boolean[] row : dp) Arrays.fill(row, false);
        
        for(int i = 0; i < n; i++) dp[i][0] = true;
        
        if (arr[0] <= sum) dp[0][arr[0]] = true;
        
        for(int i = 1; i < n; i++){
            for(int target = 1; target <= sum; target++){
                boolean exclude = dp[i-1][target];
                boolean include = false;
                if(target >= arr[i]) include = dp[i-1][target-arr[i]];
                
                dp[i][target] = include | exclude;
            }
        }
        
        return dp[n-1][sum];
    }

    public boolean canPartition(int[] nums) {
        int sum = 0;
        for(int num : nums) sum += num;

        if(sum % 2 != 0) return false;

        int target = sum/2;

        return isSubsetSum(nums, target);
    }
}